src/Lean/Compiler/LCNF/FloatLetIn.lean

```diff
@@ -0,0 +1,263 @@
+/-
+Copyright (c) 2022 Henrik Böving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Lean.Compiler.LCNF.CompilerM
+import Lean.Compiler.LCNF.FVarUtil
+import Lean.Compiler.LCNF.PassManager
+import Lean.Compiler.LCNF.Types
+
+namespace Lean.Compiler.LCNF
+
+namespace FloatLetIn
+
+/--
+The decision of the float mechanism.
+-/
+inductive Decision where
+|
+  /--
+  Push into the arm with name `name`.
+  -/
+  arm (name : Name)
+| /--
+  Push into the default arm.
+  -/
+  default
+|
+  /--
+  Dont move this declaration it is needed where it is right now.
+  -/
+  dont
+|
+  /--
+  No decision has been made yet.
+  -/
+  unknown
+deriving Hashable, BEq, Inhabited, Repr
+
+def Decision.ofAlt : Alt → Decision
+| .alt name _ _ => .arm name
+| .default _ => .default
+
+/--
+The context for `BaseFloatM`.
+-/
+structure BaseFloatContext where
+  /--
+  All the declarations that were collected in the current LCNF basic
+  block up to the current statement.
+  -/
+  decls : Array CodeDecl := #[]
+
+/--
+The state for `FloatM`
+-/
+structure FloatState where
+  /--
+  A map from identifiers of declarations to their current decision.
+  -/
+  decision : HashMap FVarId Decision
+  /--
+  A map from decisions (excluding `unknown`) to the declarations with
+  these decisions. Basically:
+  - Which declarations do we not move
+  - Which declarations do we move into a certain arm
+  - Which declarations do we move into the default arm
+  -/
+  newArms : HashMap Decision (Array CodeDecl)
```

Comment:
`HashMap`s of arrays are often problematic because they require discipline to update the arrays linearly. 
For example, in the following code
```
s.newArms.insert .dont (s.newArms.find! .dont |>.push decl)
```
we first retrieve the array from the hashmap. Now, we have two references to it: one inside of the hashmap, and another that we are passing to `Array.push`. Thus, the array is not updated destructively. 
To ensure it is, we need to remove the entry from the hashmap, then update the array, and finally reinsert it.
In the future, we hope the have nice combinators for performing this kind of operation.
Right now, I think it is easier to just use `HashMap Decision (List CodeDecl)`.